### PR TITLE
TELCODOCS#1877: Enabling the SR-IOV network metrics exporter

### PR DESCRIPTION
**Pull request by sr1kar99**
Version(s):
4.17

Issue:
[TELCODOCS-1877](https://issues.redhat.com/browse/TELCODOCS-1877)

Link to docs preview:
- [About the SR-IOV network metrics exporter](https://80754--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-sriov-operator.html#sriov-network-metrics-exporter_configuring-sriov-operator)
- [Enabling the SR-IOV network metrics exporter](https://80754--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-sriov-operator.html#sriov-operator-metrics_configuring-sriov-operator)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

**Comment by sr1kar99** [comment]
@zeeke 
Could you please review this PR?
Thanks! :)

File: modules/sriov-operator-metrics.adoc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+* You have installed the SR-IOV Network Operator.
+
+.Procedure
+
+* Set the `featureGates` field to `true` by running the following command:
++
+[source,terminal]
+----
+$ oc patch -n openshift-sriov-network-operator sriovoperatorconfig/default \
+    --type='json' -p='[{"op": "replace", "path": "/spec/featureGates/metricsExporter", "value":true}]'
```
Comment (zeeke) [inline modules/sriov-operator-metrics.adoc:39]: this does not work if the `.spec.featureGates` map is nil:

```suggestion
$ oc patch -n openshift-sriov-network-operator sriovoperatorconfig/default \
    --type='merge'  -p='{"spec": {"featureGates": {"metricsExporter": true}}}'
```

File: modules/sriov-operator-metrics.adoc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+[source,yaml]
+----
+apiVersion: sriovnetwork.openshift.io/v1
+kind: SriovOperatorConfig
+metadata:
+  name: default
+  namespace: openshift-sriov-network-operator
+spec:
+  featureGates: true
```
Comment (zeeke) [inline modules/sriov-operator-metrics.adoc:54]:
```suggestion
  featureGates: 
    metricsExporter: true
```

File: modules/nw-sriov-configuring-operator.adoc [hunk 3]
```diff
@@ -86,6 +87,11 @@ By default, this field is set to `true`.
 Set to `0` to show only the basic logs. Set to `2` to show all the available logs.
 By default, this field is set to `2`.
 
+|`spec.featureGates`
+|`boolean`
+|Specifies whether to enable or disable the optional features.
+By default, this field is set to `false`.
+
```
Comment (zeeke) [inline modules/nw-sriov-configuring-operator.adoc:94]:
```suggestion
|`spec.featureGates`
|`map[string]bool`
|Specifies whether to enable or disable the optional features, e.g. `metricsExporter.
```

**Review by zeeke — COMMENTED** [review]
Left some comments. To enable the feature, the user also needs to add `openshift.io/cluster-monitoring` label to the operator namespace:

```
oc label ns openshift-sriov-network-operator openshift.io/cluster-monitoring=true
```

can you please update either the namespace creation in the install section or the "Enabling the SR-IOV Network Operator metrics" paragraph

File: modules/sriov-operator-metrics.adoc (hunk 2 above)
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc:54]: Updated as per the suggestion.

File: modules/sriov-operator-metrics.adoc (hunk 1 above)
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc:39]: Updated as per the suggestion.

File: modules/nw-sriov-configuring-operator.adoc (hunk 3 above)
Comment (sr1kar99) [reply modules/nw-sriov-configuring-operator.adoc:94]: Updated as per the suggestion.

File: modules/sriov-operator-metrics.adoc [hunk 4]
```diff
@@ -0,0 +1,77 @@
+// Module included in the following assemblies:
+//
+// * networking/hardware_networks/configuring-sriov-operator.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="sriov-operator-metrics_{context}"]
+= Enabling the SR-IOV Network Operator metrics
+
+You can enable the SR-IOV Network Operator metrics to monitor and analyze the Operator's performance.
```
Comment (zeeke) [inline modules/sriov-operator-metrics.adoc:9]: This doesn't look accurate to me. The feature allows the user to monitor workload pod networking activity, not the operator's one.

File: modules/sriov-operator-metrics.adoc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+.Procedure
+
+* Set the `spec.featureGates.metricsExporter` field to `true` by running the following command:
++
+[source,terminal]
+----
+$ oc patch -n openshift-sriov-network-operator sriovoperatorconfig/default \
+    --type='merge' -p='{"spec": {"featureGates": {"metricsExporter": true}}}'
+----
```
Comment (zeeke) [inline modules/sriov-operator-metrics.adoc:52]: Please, add the namespace labeling step:
```
oc label ns/openshift-sriov-network-operator openshift.io/cluster-monitoring=true
```

This is needed for the feature to be activated

Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc:52]: Updated as per the suggestion.

File: modules/sriov-operator-metrics.adoc (hunk 4 above)
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc]: Updated as per the suggestion.

File: modules/nw-sriov-configuring-operator.adoc [hunk 6]
```diff
@@ -22,6 +22,7 @@ spec:
   enableInjector: true
   enableOperatorWebhook: true
   logLevel: 2
+  featureGates: false
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/nw-sriov-configuring-operator.adoc]: https://80754--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-sriov-operator.html#sriov-operator-metrics_configuring-sriov-operator

File: modules/sriov-operator-metrics.adoc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="sriov-operator-metrics_{context}"]
+= Enabling the SR-IOV network metrics exporter
+
+The SR-IOV network metrics exporter is disabled by default. To enable the metrics exporter, you must set the `spec.featureGates.metricsExporter` field to `true`.
+
+[IMPORTANT]
+====
+When the metrics exporter is enabled, the SR-IOV Network Oprator deploys the metrics exporter only on nodes that include the `feature.node.kubernetes.io/network-sriov.capable": "true"` label.
```
Comment (zeeke) [inline modules/sriov-operator-metrics.adoc:13]: Though this is probably true for most of the deployment, it's better to be more accurate: the metric-exporter is deployed on the same node as the config-daemon, which are the one described by the field `SriovOperatorConfig.Spec.ConfigDaemonNodeSelector`.

If you prefer, we can simply state something like "the metric-exporter is deployed only on node with SR-IOV capabilities"

**Review by zeeke — COMMENTED** [review]
I left one comment, beside that, it LGTM

File: modules/sriov-operator-metrics.adoc (hunk 7 above)
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc]: Updated as per the suggestion.

File: modules/sriov-network-metrics-exporter.adoc [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+You can also combine these queries with the kube-state-metrics to get more information about the SR-IOV pods. For example, you can use the following query to get the VF network statistics along with the application name from the standard Kubernetes pod label:
+[source,bash]
+----
+(sriov_vf_tx_packets * on (pciAddr)  group_left(pod,namespace)  sriov_kubepoddevice) * on (pod,namespace) group_left (label_app_kubernetes_io_name) kube_pod_labels
+----
+
+[NOTE]
+====
+To view the network statistics of a VF by using the {oc-first}, run the following command:
```
Comment (zeeke) [inline modules/sriov-network-metrics-exporter.adoc:40]: If we go down this way, we have to explain the user to open a shell in the worker node, that is not related to the `oc` client. 
This is not how the feature is supposed to be used, it's an information for the user to know where these numbers come from, as he might be used to kernel device statistics.

That said, I'm in favor of removing this note: the introduction you wrote at the beginning of this chapter is clear enough.

If you prefer to keep it, please specify that the user has to run the `ip -s link ...` command on the worker node. Your call

Comment (sr1kar99) [reply modules/sriov-network-metrics-exporter.adoc]: @zeeke 
I agree with you!
This note is just additional info. Also, the metrics have been described in the table with example so I think we can skip the note.

Updated as per the suggestion.

**Review by ajaggapa — COMMENTED** [review]
/lgtm

**Review by adellape — COMMENTED** [review]
Only a few minor questions, looks good!

File: modules/sriov-network-metrics-exporter.adoc [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+|`sriov_vf_tx_packets` |Transmitted packets per virtual function. |`sriov_vf_tx_packets * on (pciAddr) group_left(pod,namespace,dev_type) sriov_kubepoddevice`
+|`sriov_vf_rx_dropped` |Dropped packets upon receipt per virtual function. |`sriov_vf_rx_dropped * on (pciAddr) group_left(pod,namespace,dev_type) sriov_kubepoddevice`
+|`sriov_vf_tx_dropped` |Dropped packets during transmission per virtual function. |`sriov_vf_tx_dropped * on (pciAddr) group_left(pod,namespace,dev_type) sriov_kubepoddevice`
+|`sriov_vf_rx_multicast` |Received multicast packets per virtual function. |`sriov_vf_rx_multicast * on (pciAddr) group_left(pod,namespace,dev_type) sriov_kubepoddevice`
+|`sriov_vf_rx_broadcast` |Received broadcast packets per virtual function. |`sriov_vf_rx_broadcast * on (pciAddr) group_left(pod,namespace,dev_type) sriov_kubepoddevice`
+|`sriov_kubepoddevice` |Virtual functions linked to active pods. |-
+
+|====
+
+You can also combine these queries with the kube-state-metrics to get more information about the SR-IOV pods. For example, you can use the following query to get the VF network statistics along with the application name from the standard Kubernetes pod label:
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/sriov-network-metrics-exporter.adoc:32]: Yes! I noticed the same, and thought to keep it consistent with other instances in the doc.

File: modules/sriov-network-metrics-exporter.adoc [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+[source,bash]
+----
+(sriov_vf_tx_packets * on (pciAddr)  group_left(pod,namespace)  sriov_kubepoddevice) * on (pod,namespace) group_left (label_app_kubernetes_io_name) kube_pod_labels
+----
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/sriov-network-metrics-exporter.adoc:36]: Updated as per the suggestion.

File: modules/sriov-operator-metrics.adoc [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+spec:
+  featureGates:
+    metricsExporter: true
+----
+====
+
+[NOTE]
+====
+After enabling the metrics exporter, you can examine the SR-IOV virtual function (VF) metrics by using the {product-title} web console. For more information, see "Querying metrics".
+====
```
[the remark this answers is not in the document]
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc:61]: Added it as an optional verification step.

File: modules/sriov-operator-metrics.adoc [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+apiVersion: sriovnetwork.openshift.io/v1
+kind: SriovOperatorConfig
+metadata:
+  name: default
+  namespace: openshift-sriov-network-operator
+spec:
+  featureGates:
+    metricsExporter: true
+----
+====
```
Comment (jeana-redhat) [inline modules/sriov-operator-metrics.adoc:56]: We should avoid showing this type of alternative method unless it represents a substantially different use case. From what I understand, `oc patch` is preferred to manually editing YAML config files.
```suggestion
```

**Review by jeana-redhat — COMMENTED** [review]
Just a couple small things here

/remove-label merge-review-in-progress
/remove-label merge-review-needed

File: modules/sriov-operator-metrics.adoc (hunk 12 above)
Comment (sr1kar99) [reply modules/sriov-operator-metrics.adoc:56]: Updated as per the suggestion.

**Pull request merged**